gadget/update.go

```diff
@@ -247,6 +278,35 @@ func EnsureLayoutCompatibility(gadgetLayout *LaidOutVolume, diskLayout *OnDiskVo
 		}
 	}
 
+	// check all structures in the layout are present in the gadget, or have a
+	// valid excuse for absence (i.e. mbr or creatable structures at install)
+	for _, gs := range gadgetLayout.LaidOutStructure {
+		// we ignore reasonAbsent here since if there was an extra on disk
+		// structure that didn't match something in the YAML, we would have
+		// caught it above, this loop can only ever identify structures in the
+		// YAML that are not on disk at all
+		if present, _ := onDiskContains(diskLayout.Structure, gs); present {
+			continue
+		}
+
+		// otherwise not present, figure out if it has a valid excuse
+
+		// allow specifically the mbr structure to be missing from ondisk since
+		// it doesn't show up as a device since it's a "raw" thing
+		// TODO: should this use Type == mbr or Role == mbr or both ?
```

Comment:
once setImplicitForVolume was run Role should be enough

Reply:
ack, just checks Role now